File: PageTransitionAnimation/app/src/main/java/com/kuanquan/pagetransitionanimation/MainFragment.java
package com.kuanquan.pagetransitionanimation;

import android.app.SharedElementCallback;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.transition.ChangeBounds;
import android.transition.ChangeImageTransform;
import android.transition.TransitionSet;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;

import androidx.core.app.ActivityOptionsCompat;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.kuanquan.pagetransitionanimation.adapter.MyAdapter;
import com.kuanquan.pagetransitionanimation.elementspage.ShareElementsActivity;
import com.kuanquan.pagetransitionanimation.util.DataUtil;
import com.kuanquan.pagetransitionanimation.viewpager.BaseFragment;

import org.jetbrains.annotations.Nullable;

import java.io.Serializable;
import java.util.List;
import java.util.Map;

public class MainFragment  extends BaseFragment {

    public List<String> datas = DataUtil.INSTANCE.getData();
    private MyAdapter myAdapter;
//    MainActivity mainActivity;
    public GridLayoutManager gridLayoutManager;

    @Override
    protected int getLayoutId() {
        return R.layout.fragment_main;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        RecyclerView recyclerView = getView().findViewById(R.id.recycler_view);
        gridLayoutManager = new GridLayoutManager(requireActivity(), 2);
        myAdapter = new MyAdapter(getActivity(), datas);
        recyclerView.setAdapter(myAdapter);
        recyclerView.setLayoutManager(gridLayoutManager);
        myAdapter.setOnClickListener(new MyAdapter.OnClickListener() {
            @Override
            public void onClick(View v, int position) {
                Intent intent = new Intent(getContext(), ShareElementsActivity.class);
                intent.putExtra("url", (Serializable) datas);
                intent.putExtra("index", position);

                // TODO 1. 共享元素动画  必备的步骤
                Bundle options = ActivityOptionsCompat.makeSceneTransitionAnimation(requireActivity(), v, datas.get(position)).toBundle();
//                startActivityForResult(intent,200,options);


                startActivity(intent,options);
//                startActivity(intent, ActivityOptions.makeSceneTransitionAnimation(MainActivity.this, v, "sharedView").toBundle());
            }
        });


        // TODO 共享元素动画  必备的步骤
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            getActivity().setExitSharedElementCallback(new SharedElementCallback() {
                @Override
                public void onMapSharedElements(List<String> names, Map<String, View> sharedElements) {
                    if (bundle != null) {
                        int position = bundle.getInt("index", 0);
                        Log.e("onMapSharedElements", "position = " + position);
                        sharedElements.clear();
                        names.clear();
                        View itemView = gridLayoutManager.findViewByPosition(position);
                        ImageView imageView = itemView.findViewById(R.id.image);
                        names.add(itemView.getTransitionName());
                        //注意这里第二个参数，如果防止是的条目的item则动画不自然。放置对应的imageView则完美
                        sharedElements.put(datas.get(position), imageView);
                        bundle = null;
                    }
                }
            });
        }

//        TransitionSet mtransitionset=new TransitionSet();//制定过度动画set
//        mtransitionset.addTransition(new ChangeBounds());//改变表框大小
//        mtransitionset.addTransition(new ChangeImageTransform());//图片移动，还可以是其他的，要什么效果自己添加
//        mtransitionset.setDuration(250);
//
//        getActivity().getWindow().setEnterTransition(mtransitionset);//注意，下面是必须的
//        getActivity().getWindow().setExitTransition(mtransitionset);
//        getActivity().getWindow().setSharedElementEnterTransition(mtransitionset);
//        getActivity().getWindow().setSharedElementExitTransition(mtransitionset);

//        mainActivity = (MainActivity)getActivity();

//        View itemView = gridLayoutManager.findViewByPosition(position);
//        ImageView imageView = itemView.findViewById(R.id.image);
//        mainActivity.setAnimatorFragment("", imageView, datas);
    }

    public Bundle bundle;
//    @Override
//    public void onActivityResult(int requestCode, int resultCode, @androidx.annotation.Nullable Intent data) {
//        super.onActivityResult(requestCode, resultCode, data);
//        Log.e("onActivityResult", "requestCode = " + requestCode);
//        bundle = new Bundle(data.getExtras());
//    }

}

File: PageTransitionAnimation/app/src/main/java/com/kuanquan/pagetransitionanimation/MainFragment.kt
package com.kuanquan.pagetransitionanimation

import android.app.SharedElementCallback
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.ImageView
import androidx.core.app.ActivityOptionsCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.kuanquan.pagetransitionanimation.adapter.MainAdapter
import com.kuanquan.pagetransitionanimation.bean.MainData
import com.kuanquan.pagetransitionanimation.elementspage.ShareElementsActivity
import com.kuanquan.pagetransitionanimation.util.DataUtil.getMainData
import com.kuanquan.pagetransitionanimation.viewpager.BaseFragment
import java.io.Serializable

class MainFragment: BaseFragment(), MainAdapter.OnAdapterItemClickListener {

    var datas: MutableList<MainData> = getMainData()
    private val myAdapter: MainAdapter by lazy { MainAdapter(datas) }
    val linearLayoutManager: LinearLayoutManager by lazy { LinearLayoutManager(requireActivity()) }
    var bundle: Bundle? = null

    override val layoutId = R.layout.fragment_main

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        myAdapter.setOnAdapterItemClickListener(this)
        val recyclerView: RecyclerView = view!!.findViewById(R.id.recycler_view)
        recyclerView.adapter = myAdapter
        recyclerView.layoutManager = linearLayoutManager

        // TODO 2. 共享元素动画  必备的步骤
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            activity?.setExitSharedElementCallback(object : SharedElementCallback() {
                override fun onMapSharedElements(names: MutableList<String?>, sharedElements: MutableMap<String, View>) {
                    if (bundle != null) {
                        val position: Int = bundle?.getInt("index", 0) ?: 0
                        Log.e("onMapSharedElements", "position = $position")
                        sharedElements.clear()
                        names.clear()
                        val parentItemView = linearLayoutManager.findViewByPosition(parentPosition)

                       val childRecyclerView = parentItemView?.findViewById<RecyclerView>(R.id.recyclerView)

                        val itemView: View = childRecyclerView?.layoutManager?.findViewByPosition(position)!!

//                        val itemView: View = myAdapter?.gridLayoutManager?.findViewByPosition(position)!!

                        val imageView = itemView.findViewById<ImageView>(R.id.image_iv)
                        names.add(itemView.transitionName)
                        //注意这里第二个参数，如果防止是的条目的item则动画不自然。放置对应的imageView则完美
                        sharedElements[datas[parentPosition].list[position]] = imageView
                        bundle = null
                    }
                }
            })
        }
    }

    var parentPosition = 0
    override fun onClick(position: Int, view: View, parentPosition: Int) {
        this.parentPosition = parentPosition

        val intent = Intent(context, ShareElementsActivity::class.java)
        intent.putExtra("url", datas[parentPosition].list as Serializable?)
        intent.putExtra("index", position)

        // TODO 1. 共享元素动画  必备的步骤
        val options = ActivityOptionsCompat.makeSceneTransitionAnimation(requireActivity(), view, datas[parentPosition].list[position]).toBundle()
        startActivity(intent, options)
    }
}